# Laboratorium 2 #
#### Zofia Jankowska ####

In [1]:
import numpy as np
import scipy
import matplotlib
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
from sqlalchemy import create_engine
import psycopg2 as pg
import pandas as pd

db_string = "postgresql://wbauer_adb:adb2020@pgsql-196447.vipserv.org:5432/wbauer_adb"
connection = pg.connect(host='pgsql-196447.vipserv.org', port=5432, dbname='wbauer_adb', user='wbauer_adb', password='adb2020')

db = create_engine(db_string)

connection_sqlalchemy = db.connect()

**Zadanie 1.** <br>
Ile kategorii filmów mamy w wypożyczalni?

In [3]:
df_1 = pd.read_sql('select * from category',con=connection_sqlalchemy)
print(df_1)


    category_id         name         last_update
0             1       Action 2006-02-15 09:46:27
1             2    Animation 2006-02-15 09:46:27
2             3     Children 2006-02-15 09:46:27
3             4     Classics 2006-02-15 09:46:27
4             5       Comedy 2006-02-15 09:46:27
5             6  Documentary 2006-02-15 09:46:27
6             7        Drama 2006-02-15 09:46:27
7             8       Family 2006-02-15 09:46:27
8             9      Foreign 2006-02-15 09:46:27
9            10        Games 2006-02-15 09:46:27
10           11       Horror 2006-02-15 09:46:27
11           12        Music 2006-02-15 09:46:27
12           13          New 2006-02-15 09:46:27
13           14       Sci-Fi 2006-02-15 09:46:27
14           15       Sports 2006-02-15 09:46:27
15           16       Travel 2006-02-15 09:46:27


In [4]:
df_1 = pd.read_sql('select count(category_id) from category',con=connection_sqlalchemy)
print(df_1)

   count
0     16


Jest 16 kategorii filmów.

**Zadanie 2.** <br>
Wyświetl listę kategorii w kolejności alfabetycznej. <br>

In [5]:
df_2 = pd.read_sql('select name from category order by name',con=connection_sqlalchemy)
print(df_2)

           name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel


**Zadanie 3.** <br>
Znajdź najstarszy i najmłodszy film do wypożyczenia.

In [6]:
df_3 = pd.read_sql('select title, release_year from film order by release_year',con=connection_sqlalchemy)
print(df_3)

                 title  release_year
0      Chamber Italian          2006
1     Grosse Wonderful          2006
2      Airport Pollock          2006
3    Bright Encounters          2006
4     Academy Dinosaur          2006
..                 ...           ...
995     Young Language          2006
996         Youth Kick          2006
997       Zhivago Core          2006
998  Zoolander Fiction          2006
999          Zorro Ark          2006

[1000 rows x 2 columns]


Wszystkie filmy zostały wydane w 2006, więc nie ma najstarszego lub najmłodszego filmu do wypożyczenia.

**Zadanie 4.** <br>
Ile wypożyczeń odbyło się między 2005-07-01 a 2005-08-01? 

In [19]:
df_4 = pd.read_sql("SELECT COUNT(rental_date) from rental WHERE rental_date BETWEEN '2005-07-01 00:00:00' AND '2005-08-01 23:59:59'",con=connection_sqlalchemy)
print(df_4)


   count
0   7380


Między tymi datami miało miejsce 7380 wypożyczeń.

**Zadanie 5.** <br>
Ile wypożyczeń odbyło się między 2010-01-01 a 2011-02-01? <br>

In [20]:
df_4 = pd.read_sql("SELECT COUNT(rental_date) from rental WHERE rental_date BETWEEN '2010-01-01 00:00:00' AND '2011-02-01 23:59:59'",con=connection_sqlalchemy)
print(df_4)


   count
0      0


Między tymi datami nikt nic nie wypożyczył.

**Zadanie 6.** <br>
Znajdź największą płatność wypożyczenia. <br>

In [ ]:
df_4 = pd.read_sql('select max(amount) from payment',con=connection_sqlalchemy)
print(df_4)

     max
0  11.99


Maksymalna płatnośc wyniosła 11.99

**Zadanie 7.** <br>
Znajdź wszystkich klientów z Polski, Nigerii lub Bangladeszu. <br>

In [38]:
df_7 = pd.read_sql("SELECT customer_id, first_name, last_name, city, country from (((country inner join city on country.country_id = city.country_id) inner join address on address.city_id = city.city_id) inner join customer on customer.address_id = address.address_id) where country in ('Poland', 'Nigeria', 'Bangladesh')",con=connection_sqlalchemy)
print(df_7)


    customer_id first_name  last_name             city     country
0           284      Sonia    Gregory       Benin City     Nigeria
1           318      Brian      Wyman        Bydgoszcz      Poland
2           572     Sidney   Burleson      Czestochowa      Poland
3           409     Rodney    Moeller        Deba Habe     Nigeria
4           332    Stephen     Qualls            Dhaka  Bangladesh
5           286      Velma      Lucas     Effon-Alaiye     Nigeria
6            80    Marilyn       Ross              Ife     Nigeria
7           198      Elsie     Kelley           Ikerre     Nigeria
8           103     Gladys   Hamilton           Ilorin     Nigeria
9           329      Frank   Waggoner         Jamalpur  Bangladesh
10          128   Marjorie     Tucker  Jastrzebie-Zdrj      Poland
11           18      Carol     Garcia           Kaduna     Nigeria
12          270       Leah     Curtis           Kalisz      Poland
13          501      Ruben      Geary           Lublin      Po

Jest 24 klientów z Polski, Nigerii lub Bangladeszu.

**Zadanie 8.** <br>
Gdzie mieszkają członkowie personelu? <br>

In [42]:
df8 = pd.read_sql("SELECT address, city, country from (((country inner join city on country.country_id = city.country_id) inner join address on address.city_id = city.city_id) inner join staff on staff.address_id = address.address_id)",con=connection_sqlalchemy)
print(df8)


                address        city    country
0     23 Workhaven Lane  Lethbridge     Canada
1  1411 Lillydale Drive   Woodridge  Australia


**Zadanie 9.** <br>
Ilu pracowników mieszka w Argentynie lub Hiszpanii? <br>

In [44]:
df_9 = pd.read_sql("SELECT count(country) from (((country inner join city on country.country_id = city.country_id) inner join address on address.city_id = city.city_id) inner join staff on staff.address_id = address.address_id) where country in ('Argentina', 'Spain')",con=connection_sqlalchemy)
print(df_9)

   count
0      0


Żaden pracownik nie mieszka w Argentynie lub Hiszpanii.

**Zadanie 10.** <br>
Jakie kategorie filmów zostały wypożyczone przez klientów? <br>

In [48]:
df_10 = pd.read_sql("SELECT distinct category.name from (((category inner join film_category on category.category_id = film_category.category_id) inner join inventory on film_category.film_id = inventory.film_id) inner join rental on inventory.inventory_id=rental.inventory_id)",con=connection_sqlalchemy)
print(df_10)

           name
0        Sports
1      Classics
2           New
3        Family
4        Comedy
5     Animation
6        Travel
7         Music
8         Drama
9        Horror
10       Sci-Fi
11        Games
12  Documentary
13      Foreign
14       Action
15     Children


Filmy we wszystkich kategoriach zostały wypożyczone.

**Zadanie 11.** <br>
Znajdź wszystkie kategorie filmów wypożyczonych w Ameryce. <br>

In [75]:
txt = "select count(name) from ((((((((category inner join film_category on film_category.category_id = category.category_id) inner join inventory on inventory.film_id = film_category.film_id) inner join rental on rental.inventory_id = inventory.inventory_id) inner join customer on rental.customer_id = customer.customer_id) inner join store on store.store_id = customer.store_id) inner join address on address.address_id = store.address_id) inner join city on city.city_id = address.city_id) inner join country on country.country_id = city.country_id) where country = 'United States'"
df_11 = pd.read_sql(txt ,con=connection_sqlalchemy)
print(df_11)

   count
0      0


Zadna kategoria nie została wypożyczona w USA.

**Zadanie 12.** <br>
Znajdź wszystkie tytuły filmów, w których grał: Olympia Pfeiffer lub Julia  Zellweger lub Ellen Presley <br>

In [61]:
df_12 = pd.read_sql("select title from ((film inner join film_actor on film_actor.film_id = film.film_id) inner join actor on actor.actor_id = film_actor.actor_id) WHERE ((first_name = 'Olympia' and last_name = 'Pfeiffer') or (first_name = 'Julia' and last_name = 'Zellweger') or (first_name = 'Ellen' or last_name = 'Presley'))",con=connection_sqlalchemy)
print(df_12)

                title
0     Bilko Anonymous
1   Caribbean Liberty
2    Casper Dragonfly
3    Empire Malkovich
4       Floats Garden
..                ...
64       Outlaw Hanky
65    Panky Submarine
66   Rider Caddyshack
67          Won Dares
68      Wyoming Storm

[69 rows x 1 columns]
